### ISSUE TITLE
TensorBoard

### ISSUE THREAD

**Issue by junxnone** [junxnone/examples#25]
#22 
> TensorBoard 是通过一些操作（summary operations）将数据记录到文件（event files）中，然后再读取文件来完成作图
# Reference
- [ ] [Tensorboard docs](https://github.com/tensorflow/tensorboard/blob/master/docs/overview.md)
- [ ] [TensorBoard:可视化学习 - 极客学院](http://wiki.jikexueyuan.com/project/tensorflow-zh/how_tos/summaries_and_tensorboard.html)
- [ ] [Understanding TensorBoard](https://github.com/secsilm/understanding-tensorboard)
- [ ] [TensorBoard可视化](https://zhuanlan.zhihu.com/p/33178205)
- [ ] [深入Tensorflow Embedding，解决Tensorboard无法显示PROJECTOR问题](https://blog.csdn.net/LiGuang923/article/details/83834431)

# Example

```
git clone https://github.com/tensorflow/tensorflow.git
cd tensorflow/tensorflow/examples/tutorials/mnist
python3 mnist_with_summaries.py
tensorboard --logdir=/tmp/tensorflow/mnist
```
Now the TensorBoard is active: http://localhost:6006/ 
![image](https://user-images.githubusercontent.com/2216970/52909533-e5afe700-32c4-11e9-9727-b433f44b8a98.png)

- Step：默认选项，指的是横轴显示的是训练迭代次数 
- Relative：这个相对指的是相对时间，相对于训练开始的时间，也就是说是训练用时 ，单位是小时 
- Time：指训练的绝对时间 

![image](https://user-images.githubusercontent.com/2216970/52909892-21997b00-32ca-11e9-86ff-a02823810fc9.png)

**Horizontal Axis 是横轴的设置**
- STEP：默认选项，指的是横轴显示的是训练迭代次数
- RELATIVE：这个相对指的是相对时间，相对于训练开始的时间，也就是说是训练用时 ，单位是小时
- WALL：指训练的绝对时间

# OVERVIEW

- summary: 在定义计算图的时候，在适当的位置加上一些 summary 操作 。
- merge: 你很可能加了很多 summary 操作，我们需要使用 tf.summary.merge_all 来将这些 summary 操作聚合成一个操作，由它来产生所有 summary 数据。
- run: 在没有运行的时候这些操作是不会执行任何东西的，仅仅是定义了一下而已。在运行（开始训练）的时候，我们需要通过 tf.summary.FileWriter() 指定一个目录来告诉程序把产生的文件放到哪。然后在运行的时候使用 add_summary() 来将某一步的 summary 数据记录到文件中。

Tensorboard栏目 | tensorflow日志生成函数 | 内容
-- | -- | --
GRAPHS | 默认保存 | 显示tensorflow计算图
SCALARS | tf.summary.scalar | 显示tensorflow中的张量随迭代轮数的变化趋势
DISTRIBUTIONS | tf.summary.histogram | 显示tensorflow中张量的直方图
HISTOGRAMS | tf.summary.histogram | 显示tensorflow中张量的直方图（以另一种方式）
IMAGES | tf.summary.image | 显示tensorflow中使用的图片
AUDIO | tf.summary.audio | 显示tensorflow中使用的音频
TEXT | tf.summary.text | 显示tensor flow中使用的文本
PROJECTOR |   | 通过读取checkpoint文件可视化高维数据
PR CURVES |  | 
PROFILE |  |  
BEHOLDER |  |  

## SCALARS
> SCALARS 主要用于记录诸如准确率、损失和学习率等单个值的变化趋势。

```
with tf.name_scope('accuracy'):
    correct_prediction = tf.equal(tf.argmax(logits, 1), tf.argmax(y, 1))
    accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
    tf.summary.scalar('accuracy', accuracy)
```
`tf.summary.scalar(name, tensor) `有两个参数：

- name：可以理解为图的标题。在 GRAPHS 中则是该节点的名字
- tensor：包含单个值的 tensor，说白了就是作图的时候要用的数据

**常用的值**
- accuracy
- loss
- eval_accuracy

- 

## IMAGES
> 如果你的模型输入是图像（的像素值），然后你想看看模型每次的输入图像是什么样的，以保证每次输入的图像没有问题（因为你可能在模型中对图像做了某种变换，而这种变换是很容易出问题的），IMAGES 面板就是干这个的，它可以显示出相应的输入图像，默认显示最新的输入图像

## GRAPHS
> 很多时候我们的模型很复杂，包含很多层，我们想要总体上看下构建的网络到底是什么样的，这时候就用到 GRAPHS 面板了，在这里可以展示出你所构建的网络整体结构，显示数据流的方向和大小，也可以显示训练时每个节点的用时、耗费的内存大小以及参数多少。默认显示的图分为两部分：主图（Main Graph）和辅助节点（Auxiliary Nodes）。其中主图显示的就是网络结构，辅助节点则显示的是初始化、训练、保存等节点

![image](https://user-images.githubusercontent.com/2216970/52909877-f6169080-32c9-11e9-9643-490eef478ca5.png)

TensorBoard 默认是不会记录每个节点的用时、耗费的内存大小等这些信息的，那么如何才能在图上显示这些信息呢？关键就是如下这些代码，主要就是在 sess.run() 中加入 options 和 run_metadata 参数。
```
run_options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
run_metadata = tf.RunMetadata()
s, lss, acc , _ = sess.run([merged_summary, loss, accuracy, train_step], 
                           feed_dict={x: batch_x, y: batch_y, phase: 1},
                           options=run_options,
                           run_metadata=run_metadata)
summary_writer.add_run_metadata(run_metadata, 'step{}'.format(i))
summary_writer.add_summary(s, i)
``` 

**Device**
![image](https://user-images.githubusercontent.com/2216970/52909845-915b3600-32c9-11e9-9feb-8af9c21fbba6.png)

**Memory & Compute Time**
![image](https://user-images.githubusercontent.com/2216970/52909823-56f19900-32c9-11e9-9366-bbabac6188eb.png)

## DISTRIBUTIONS
> 主要用来展示网络中各参数随训练步数的增加的变化情况，可以说是 多分位数折线图 的堆叠。

图表上的每条线表示数据分布的百分位数，例如，底线显示最小值随时间的变化趋势，中间的线显示中值变化的方式。从上至下看时，各行具有以下含义：[最大值，93％，84％，69％，50％，31％，16％，7％，最小值]。这些百分位数也可以看作标准偏差的正态分
![image](https://user-images.githubusercontent.com/2216970/52909962-91f4cc00-32cb-11e9-9008-148e47b0323b.png)

## HISTOGRAMS
> 与 DISTRIBUTIONS 不同的是，HISTOGRAMS 可以说是 频数分布直方图 的堆叠。

![image](https://user-images.githubusercontent.com/2216970/52909916-ad130c00-32ca-11e9-9087-74f324198ca1.png)

- OFFSET 横轴表示权重值，纵轴表示训练步数。默认为 OFFSET 模式。
- OVERLAY 时横轴为权重值，纵轴为频数，每一条折线为训练步数。颜色深浅与上面同理。

 ## PROJECTOR
## PROFILE
